internal/types/action/registrar.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package action
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/boundary/internal/types/resource"
+)
+
+type resourceActionSets struct {
+	individual ActionSet
+	collection ActionSet
+	valid      ActionSet
+}
+
+type byResource struct {
+	mu sync.RWMutex
+	m  map[resource.Type]*resourceActionSets
+}
+
+func (b *byResource) add(r resource.Type, individual ActionSet, collection ActionSet) error {
```

Comment:
should this just panic if there's an error since it's only used during
registration which should only happen during init?

Reply:
I updated RegisterResource to panic if called twice with the same resource, see: 21f7ac863. And then I refactored where this is called with the fixup here: 0cb15e6ec.